packages/vite/src/node/server/transformRequest.ts

```diff
@@ -419,37 +438,40 @@ async function loadAndTransform(
  * - SSR: We don't need to change anything as `ssrLoadModule` controls it
  */
 async function handleModuleSoftInvalidation(
-  mod: ModuleNode,
-  ssr: boolean,
+  environment: DevEnvironment,
+  mod: EnvironmentModuleNode,
   timestamp: number,
-  server: ViteDevServer,
 ) {
-  const transformResult = ssr ? mod.ssrInvalidationState : mod.invalidationState
+  const transformResult = mod.invalidationState
 
   // Reset invalidation state
-  if (ssr) mod.ssrInvalidationState = undefined
-  else mod.invalidationState = undefined
+  mod.invalidationState = undefined
 
   // Skip if not soft-invalidated
   if (!transformResult || transformResult === 'HARD_INVALIDATED') return
 
-  if (ssr ? mod.ssrTransformResult : mod.transformResult) {
+  if (mod.transformResult) {
     throw new Error(
       `Internal server error: Soft-invalidated module "${mod.url}" should not have existing transform result`,
     )
   }
 
   let result: TransformResult
-  // For SSR soft-invalidation, no transformation is needed
-  if (ssr) {
+  // No transformation is needed if it's disabled manually
+  // This is primarily for backwards compatible SSR
+  if (!environment.config.injectInvalidationTimestamp) {
     result = transformResult
   }
-  // For client soft-invalidation, we need to transform each imports with new timestamps if available
+  // We need to transform each imports with new timestamps if available
   else {
-    await init
     const source = transformResult.code
     const s = new MagicString(source)
-    const [imports] = parseImports(source, mod.id || undefined)
+    const imports = transformResult.ssr
+      ? await ssrParseImports(mod.url, source)
+      : await (async () => {
+          await init
+          return parseImports(source, mod.id || undefined)[0]
+        })()
```

Comment:
Why do we need a different function to parse imports in SSR?

Reply:
SSR code uses `__vite_import__`

Reply:
Got it. It would be nice it we use a regex to capture instead of AST though.